internal/wal/wal.go

```diff
@@ -28,28 +45,131 @@ type WalOptions struct {
 	Seed           uint64
 	Path           string
 	MaxSegmentSize int
+	readOnly       bool
+}
+
+func (opt WalOptions) IsValid() bool {
+	return len(opt.Path) > 0 && opt.MaxSegmentSize >= MinSegmentSize
 }
 
 type Wal struct {
 	opts   WalOptions
 	active *segment
 	csum   uint64
 	hash   hash.Hash64
+	sz     int64
+	lock   sync.Mutex
 }
 
 func Create(opts WalOptions) (*Wal, error) {
+	if !opts.IsValid() {
+		return nil, ErrInvalidWalOption
+	}
 	// create directory
+	err := os.MkdirAll(opts.Path, 0750)
+	if err != nil && !errors.Is(err, os.ErrExist) {
+		return nil, err
+	}
 	// create active wal segment
-	return &Wal{}, nil
+	// use the seed as first checksum
+	id := NewLSN(0, int64(opts.MaxSegmentSize), 0)
+	seg, err := createSegment(id, opts)
+	if err != nil {
+		return nil, err
+	}
+	return &Wal{
+		opts:   opts,
+		active: seg,
+		hash:   xxhash.New(),
+		csum:   opts.Seed,
+	}, nil
 }
 
-func Open(lsn LSN, opts WalOptions) (*Wal, error) {
+func Open(id LSN, opts WalOptions) (*Wal, error) {
+	if !opts.IsValid() {
+		return nil, ErrInvalidWalOption
+	}
 	// try open directory
 	// set exclusive lock
 	// open last segment file
 	// read hash of last record and init w.csum
+	seg, err := openSegment(id, opts)
+	if err != nil {
+		return nil, err
+	}
 
-	return &Wal{}, nil
+	// if we have been able to load the segment LSN
+	// we can assume we the minimum size
+	// of the wal is LSN - 1
+	wal := &Wal{
+		opts:   opts,
+		active: seg,
+		hash:   xxhash.New(),
+		csum:   opts.Seed,
+		sz:     int64(id) - 1,
+	}
+	var lastValidRecord, record *Record
+	r := wal.NewReader()
+	defer r.Close()
+	for {
+		record, err = r.Next()
+		if err != nil {
+			var f *os.File
+			switch {
+			case errors.Is(err, ErrChecksum):
+				name := generateFilename(int64(r.ReadSegmentId()))
```

Comment:
Refactor this into a separate function `wal.Truncate(LSN) error`. Track the LSN of the current record because `r.ReadPosition()` gives you the wrong info when you hit a checksum error. Also close the reader BEFORE truncate to avoid trouble reopening the same file.